pallets/pallet-asset-swap/src/lib.rs

```diff
@@ -0,0 +1,548 @@
+				return Err(DispatchError::from(Error::<T>::Internal));
+			}
+
+			// 9. Send XCM out
+			T::XcmRouter::deliver(xcm_ticket.0).map_err(|e| {
+				log::info!("Failed to deliver ticket with error {:?}", e);
+				DispatchError::from(Error::<T>::Xcm)
+			})?;
+
+			// 10. Update remote asset balance
+			SwapPair::<T>::try_mutate(|entry| {
+				let Some(SwapPairInfoOf::<T> {
+					remote_asset_balance, ..
+				}) = entry.as_mut()
+				else {
+					log::error!(target: LOG_TARGET, "Failed to borrow stored swap pair info as mut.");
+					return Err(Error::<T>::Internal);
+				};
+				let Some(new_balance) = remote_asset_balance.checked_sub(remote_asset_amount_as_u128) else {
+					log::error!(target: LOG_TARGET, "Failed to subtract {:?} from stored remote balance {:?}.", transferred_amount, remote_asset_balance);
+					return Err(Error::<T>::Internal);
+				};
+				*remote_asset_balance = new_balance;
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::LocalToRemoteSwapExecuted {
+				from: submitter,
+				to: *beneficiary,
+				amount: local_asset_amount,
+			});
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn set_swap_pair_bypass_checks(
+		reserve_location: VersionedMultiLocation,
+		remote_asset_id: VersionedAssetId,
+		remote_fee: VersionedMultiAsset,
+		total_issuance: u128,
+		circulating_supply: u128,
+		pool_account: T::AccountId,
+	) {
+		debug_assert!(
+			total_issuance >= circulating_supply,
+			"Provided total issuance smaller than circulating supply."
+		);
+		let swap_pair_info = SwapPairInfoOf::<T> {
+			pool_account: pool_account.clone(),
+			// We can do a simple subtraction since all checks are performed in calling functions.
+			remote_asset_balance: total_issuance - circulating_supply,
+			remote_asset_id: remote_asset_id.clone(),
+			remote_fee: remote_fee.clone(),
+			remote_reserve_location: reserve_location.clone(),
+			status: SwapPairStatus::Paused,
+		};
+
+		SwapPair::<T>::set(Some(swap_pair_info));
+
+		Self::deposit_event(Event::<T>::SwapPairCreated {
+			circulating_supply,
+			pool_account,
+			remote_asset_reserve_location: reserve_location,
+			remote_asset_id,
+			remote_xcm_fee: Box::new(remote_fee),
+			total_issuance,
+		});
+	}
+
+	fn unset_swap_pair_bypass_checks() {
+		let swap_pair = SwapPair::<T>::take();
+		if let Some(swap_pair) = swap_pair {
+			Self::deposit_event(Event::<T>::SwapPairRemoved {
+				remote_asset_id: swap_pair.remote_asset_id,
+			});
+		};
+	}
+
+	fn set_swap_pair_status(new_status: SwapPairStatus) -> Result<(), Error<T>> {
+		SwapPair::<T>::try_mutate(|entry| {
+			let SwapPairInfoOf::<T> {
+				remote_asset_id,
+				status,
+				..
+			} = entry.as_mut().ok_or(Error::<T>::SwapPairNotFound)?;
+			let relevant_event = match new_status {
+				SwapPairStatus::Running => Event::<T>::SwapPairResumed {
+					remote_asset_id: remote_asset_id.clone(),
+				},
+				SwapPairStatus::Paused => Event::<T>::SwapPairPaused {
+					remote_asset_id: remote_asset_id.clone(),
+				},
```

Comment:
A bit picky, but we should also add the remote_reserve_location in those events.

Reply:
Also, the pallet supports a single pair, so there's not even confusion for any external indexers. I really don't see any benefits in also including the remote location here, but I'm open to change my mind.

Reply:
I was assuming, once there are multiple swapPairs, it could get ambiguous, or is it out of discussion and we only support a single swapPair?

Reply:
If we would support multiple pairs within the same pallet, then we would probably add the local and remote asset IDs in the event, but still for the reserve location there would be no need, since it cannot change afterwards anyway. Indexers would probably stored it the first time they see it, and assume that will stay like that.

Reply:
Yes, sorry. That's true. I was assuming the smart contract could be deployed on multiple EVM based chains, but the remote asset IDs would be different every time.